compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/types/Types.java

```diff
@@ -0,0 +1,201 @@
+        this.ERROR = typesFactory.getTypeDescriptor(symbolTable.errorType);
+        this.FUNCTION = typesFactory.getTypeDescriptor(symbolTable.invokableType);
+        this.FUTURE = typesFactory.getTypeDescriptor(symbolTable.futureType);
+        this.TYPEDESC = typesFactory.getTypeDescriptor(symbolTable.typeDesc);
+        this.HANDLE = typesFactory.getTypeDescriptor(symbolTable.handleType);
+        this.STREAM = typesFactory.getTypeDescriptor(symbolTable.streamType);
+        this.ANY = typesFactory.getTypeDescriptor(symbolTable.anyType);
+        this.ANYDATA = typesFactory.getTypeDescriptor(symbolTable.anydataType);
+        this.NEVER = typesFactory.getTypeDescriptor(symbolTable.neverType);
+        this.READONLY = typesFactory.getTypeDescriptor(symbolTable.readonlyType);
+        this.JSON = typesFactory.getTypeDescriptor(symbolTable.jsonType);
+        this.BYTE = typesFactory.getTypeDescriptor(symbolTable.byteType);
+        this.COMPILATION_ERROR = typesFactory.getTypeDescriptor(symbolTable.semanticError);
+    }
+
+    /**
+     * Lookup for the symbol of a user defined type within a given module. This would be considering type
+     * definitions, enums, enum members, and class definitions as valid user defined types when
+     * looking up. The module is determined by the provided org, module name and the version.
+     *
+     * @param org           The organization of the looking up module
+     * @param moduleName    The name of the looking up module
+     * @param version       The version of the looking up module
+     * @param typeDefName   The type definition
+     * @return The {@link Symbol} of the user defined type
+     */
+    public Optional<Symbol> getTypeByName(String org, String moduleName, String version, String typeDefName) {
+        PackageID packageID = new PackageID(Names.fromString(org), Names.fromString(moduleName),
+                Names.fromString(version));
+
+        return getTypeDefByName(packageID, typeDefName);
+    }
+
+    /**
+     * Lookup for all the symbols of user defined types within a given module. This would be considering type
+     * definitions, enums, enum members, and class definitions as valid user defined types when looking up.
+     * The module is determined by the provided org, module name and the version.
+     *
+     * @param org           The organization of the looking up module
+     * @param moduleName    The name of the looking up module
+     * @param version       The version of the looking up module
+     * @return A {@link Map} of the user defined type symbols
+     */
+    public Optional<Map<String, Symbol>> typesInModule(String org, String moduleName, String version) {
+        PackageID packageID = new PackageID(Names.fromString(org), Names.fromString(moduleName),
+                Names.fromString(version));
+
+        return getTypeDefSymbolsInModule(packageID);
+    }
+
+    public static Types getInstance(CompilerContext context) {
```

Comment:
Add a doc comment to this clearly mentioning that this is not a part of the public API and not intended for external users' use.

Reply:
If possible, would be even better if we can figure out a way to truly restrict external users' attempts to use this. 🤔

Reply:
A much better solution would be to use `Types.Java` as an interface that exposes all the public APIs and make this the implementation of it.

Reply:
Hmm. I suppose then the tradeoff would be the use of the builtin types as `types.int()` instead of `Types.INT`

Reply:
@pubudu91 used an abstract class as the public interface to access the Types API. This way, it was able to preserve the design of the built-in types naming convention.